engine/runtime-integration-tests/src/test/java/org/enso/interpreter/test/instrument/DebugServerWithScriptTest.java

```diff
@@ -0,0 +1,106 @@
+import org.enso.common.DebugServerInfo;
+import org.enso.test.utils.ContextUtils;
+import org.graalvm.polyglot.Context;
+import org.hamcrest.core.AllOf;
+import org.junit.AfterClass;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+public class DebugServerWithScriptTest {
+  private static Context ctx;
+  private static ByteArrayOutputStream out = new ByteArrayOutputStream();
+  private static ByteArrayOutputStream err = new ByteArrayOutputStream();
+
+  @BeforeClass
+  public static void initContext() throws Exception {
+    var b = ContextUtils.defaultContextBuilder().out(out).err(err);
+    b.option(DebugServerInfo.FN_OPTION, "ScriptTest.inspect");
+    ctx = b.build();
+  }
+
+  @AfterClass
+  public static void closeContext() throws Exception {
+    ctx.close();
+    ctx = null;
+    out = null;
+    err = null;
+  }
+
+  @Before
+  public void cleanSteams() {
+    out.reset();
+    err.reset();
+  }
+
+  @Test
+  public void listingVariablesWithWarnings() throws Exception {
+    var code =
+        """
+        from Standard.Base import all
+
+        inspect =
+            j = 1
+            d = Warning.attach "doubled value" 2
+            t = j + d
+            v = [j, d, t]
+            v
+        """;
+    var r = ContextUtils.evalModule(ctx, code, "ScriptTest.enso", "inspect");
+    assertTrue("Got array back: " + r, r.hasArrayElements());
+    assertEquals("Got three elements", 3, r.getArraySize());
+    assertEquals("One", 1, r.getArrayElement(0).asInt());
+    assertEquals("Two", 2, r.getArrayElement(1).asInt());
+    assertEquals("Three", 3, r.getArrayElement(2).asInt());
+    assertEquals("No output printed", "", out.toString());
+    assertThat(
+        "Stderr contains some warnings",
+        err.toString(),
+        AllOf.allOf(
+            containsString("d = 2"),
+            containsString("t = 3"),
+            containsString("doubled value"),
+            not(containsString("j = 1"))));
+
+    var at1 = err.toString().indexOf("d = 2");
+    assertNotEquals("d = 2 found", -1, at1);
+    var at2 = err.toString().indexOf("d = 2", at1 + 1);
+    assertEquals("d = 2 not found for the second time", -1, at2);
+  }
+
+  @Test
+  public void panicOnError() throws Exception {
+    var code =
+        """
+        from Standard.Base import all
+
+        inspect =
+            j = 1
+            d = Error.throw 2
+            t = j + d
+            v = [j, d, t]
+            v
+        """;
+    var r = ContextUtils.evalModule(ctx, code, "ScriptTest.enso", "inspect");
+    assertTrue("Got error back: " + r, r.isException());
+    assertEquals("(Error: 2)", r.toString());
+    assertEquals("No output printed", "", out.toString());
+    assertThat(
+        "Stderr contains some errors",
+        err.toString(),
+        AllOf.allOf(
+            containsString("d = Error:2"),
+            containsString("t = Error:2"),
+            not(containsString("j = 1"))));
+  }
```

Comment:
How does the new error propagation work?

What will happen if I run this program?
```
main =
    b = True
    e = Error.throw 123
    r = if b then "OK" else e
    r
```

If _any_ binding in the `main` function returns a dataflow error, we break from the function with error? Does that mean that the main function has different semantics from **every other function**? Or am I misunderstanding something? If yes then this feels dangerous - shouldn't semantics of error propagations be the same everywhere? Maybe we could solve this instead by implementing something like #5430?

Reply:
I'd propose to add the above snippet as a unit test, to show the behaviour in yet another case.

The 'problem' with the snippet that is being tested already is that there **is _some_** dataflow dependency between the result `v` and the errored statement `d`. The snippet is OK, but it does not show some edge cases (the situation where the result does not depend on the error).

Possibly we could even have a test for this as well:
```
main =
    unused = Error.throw 123
    r = 123
    r
```

Reply:
> Does that mean that the main function has different semantics from **every other function**?

- I believe the whole purpose of #9749 is to give `main` function different semantics

Reply:
I just had a chat with @jdunkerley and he was not aware that we are changing the **return value** of `main`.

The idea was to:
1. report warnings and errors - this is a change that does not change semantics in a significant way if we treat these reports as diagnostic logs,
2. change the exit code of the runner to failure if any error was encountered - this is indeed a bit of a semantics change, but it is just affecting the exit code - not the return value.

Admittedly, the _return value of `main`_ isn't currently really used anywhere else apart from being printed at the end, so this is a very subtle difference. But I'm worried that this change of semantics would stay with us for too long.

Overall the agreement with @jdunkerley is that the exit code should be set to failure if an error is encountered, but the result of the `main` function should _not_ be affected. Can we do that?

Reply:
> I'd propose to add the above snippet as a unit test, to show the behaviour in yet another case.

Test added in 8f6ebdb

Reply:
> code should be set to failure if an error is encountered, but the result of the `main` function should _not_ be affected. Can we do that?

I managed to do it in 47c4e684da. Now I am more satisfied with this change and I believe you and James will be too.